### PR TITLE
3927 odk media processing

### PR DESCRIPTION
**Pull request by meltheadorable**

### CODE REVIEW

File: app/controllers/concerns/odk_headerable.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+module OdkHeaderable
+  extend ActiveSupport::Concern
+
+  included do
+    skip_authorize_resource only: :odk_headers
+    before_action :add_openrosa_headers
+
+    def odk_headers
```
Comment (hooverlunch) [inline app/controllers/concerns/odk_headerable.rb]: Why is this method in the `included` block? Isn't it just an instance method?

File: app/controllers/concerns/odk_headerable.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      authorize! :create, Response
+      # For HEAD requests, ODK wants a 204 code with a Location header for some strange reason.
+      render(nothing: true, status: 204)
+    end
+  end
+
+  # adds the appropriate headers for openrosa content
+  def add_openrosa_headers
```
Comment (hooverlunch) [inline app/controllers/concerns/odk_headerable.rb]: Should this be private?

File: app/controllers/concerns/odk_headerable.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+module OdkHeaderable
+  extend ActiveSupport::Concern
+
+  included do
+    skip_authorize_resource only: :odk_headers
+    before_action :add_openrosa_headers
+
+    def odk_headers
+      authorize! :create, Response
+      # For HEAD requests, ODK wants a 204 code with a Location header for some strange reason.
```
Comment (hooverlunch) [inline app/controllers/concerns/odk_headerable.rb]: This comment is in the wrong place I think.

File: app/controllers/concerns/odk_headerable.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      render(nothing: true, status: 204)
+    end
+  end
+
+  # adds the appropriate headers for openrosa content
+  def add_openrosa_headers
+    response.content_type = "text/xml" if request.format.xml?
+    response.headers["Location"] = request.original_url if request.method == "HEAD"
+    response.headers["X-OpenRosa-Version"] = "1.0"
+  end
```
Comment (hooverlunch) [inline app/controllers/concerns/odk_headerable.rb]: Have these changes been tested with actual submissions from the app? Just want to make sure we haven't broken anything.

File: app/controllers/forms_controller.rb [hunk 5]
```diff
@@ -100,7 +101,7 @@ def odk_manifest
       @ifa = ItemsetsFormAttachment.new(form: @form)
       @ifa.ensure_generated
     end
-    render_openrosa
+    render
```
Comment (hooverlunch) [inline app/controllers/forms_controller.rb]: Are all these render calls necessary? Aren't they implicit?

Comment (meltheadorable) [inline app/controllers/forms_controller.rb]: Probably, I think that was the result of a find/replace which would've hit places where they may not have been explicit as well, I'll clean that up.

File: app/models/media/object.rb [hunk 6]
```diff
@@ -4,11 +4,14 @@ class Media::Object < ActiveRecord::Base
...
   validates_attachment_presence :item
 
-  delegate :url, to: :item
   delegate :mission, to: :answer
 
   scope :expired, -> { where(answer_id: nil).where('created_at < ?', 12.hours.ago) }
 
+  def url(style = nil)
+    item.url(style) if item.present?
+  end
```
Comment (hooverlunch) [inline app/models/media/object.rb]: Could you just add `allow_nil: true` to the delegator?

File: app/models/xml_submission.rb [hunk 7]
```diff
@@ -50,15 +82,71 @@ def populate_from_odk(xml)
...
+    end
+  end
+
+  def generate_odk_hash
+    @odk_hash ||= Digest::SHA256.base64digest @data
+  end
+
+  def check_for_existing_response
+    response = Response.find_by(odk_hash: generate_odk_hash)
+    @response = response if response.present?
```
Comment (hooverlunch) [inline app/models/xml_submission.rb]: Very nice

File: app/models/xml_submission.rb [hunk 8]
```diff
@@ -50,15 +82,71 @@ def populate_from_odk(xml)
...
+    if id_or_str =~ /\Aon(\d+)\z/
+      # look up inputs of the form "on####" as option node ids
+      OptionNode.id_to_option_id($1)
+    else
+      # look up other inputs as option ids
+      Option.where(id: id_or_str).pluck(:id).first
+    end
+  end
+
+  def generate_odk_hash
```
Comment (hooverlunch) [inline app/models/xml_submission.rb]: I think by convention I'd call this just `odk_hash`.

File: app/controllers/concerns/odk_headerable.rb (hunk 1 above)
Comment (meltheadorable) [inline app/controllers/concerns/odk_headerable.rb]: It's a controller action, for ODK's HEAD and GET requests, I removed the conditionals from the CREATE method, so the routes now respond to HEAD/GET by directing them to `controller#odk_headers`

File: app/models/xml_submission.rb [hunk 9]
```diff
@@ -1,18 +1,51 @@
...
     @response = response
-    @data = data
     @response.source = source
     case source
-    when 'odk'
+    when "odk"
+      @data = files.delete(:xml_submission_file).read
+      @files = files
+      check_for_existing_response
+      @response.odk_hash = generate_odk_hash if @response.incomplete
```
Comment (hooverlunch) [inline app/models/xml_submission.rb]: `incomplete?`?

File: app/models/xml_submission.rb [hunk 10]
```diff
@@ -50,15 +82,71 @@ def populate_from_odk(xml)
...
+      answer.media_object = Media::Image.create(item: @files[str].open) if @files[str]
+    when "audio"
+      answer.media_object = Media::Audio.create(item: @files[str].open) if @files[str]
+    when "video"
+      answer.media_object = Media::Video.create(item: @files[str].open) if @files[str]
+    else
+      answer.value = str
+    end
+    # nullify answers if string suggests multimedia answer but no file present to make multi-chunk submissions work
+    answer = nil if (question_type.multimedia? && answer.media_object.blank?)
```
Comment (hooverlunch) [inline app/models/xml_submission.rb]: Why assign nil to answer here? Perhaps just return nil?

File: app/controllers/concerns/odk_headerable.rb (hunk 4 above)
Comment (meltheadorable) [inline app/controllers/concerns/odk_headerable.rb]: I can test them again, but I'm about 98% sure I tested with app submissions after making these changes.

File: db/migrate/20160329064209_add_odk_hash_to_responses.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+class AddOdkHashToResponses < ActiveRecord::Migration
+  def change
+    add_column :responses, :odk_hash, :string
+    add_index :responses, :odk_hash, unique: true
```
Comment (hooverlunch) [inline db/migrate/20160329064209_add_odk_hash_to_responses.rb]: Is it a good idea to make this unique? Let's discuss.

File: spec/requests/media/odk_response_spec.rb [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        expect(answer.media_object).to be_present
+      end
+    end
+  end
+
+  context "with multiple parts" do
+    before do
+      @form = create(:form, question_types: %w(image sketch))
+      @form.publish!
+      @user = create(:user,role_name: "observer")
```
Comment (hooverlunch) [inline spec/requests/media/odk_response_spec.rb]: Missing space.

File: spec/requests/media/odk_response_spec.rb [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      expect(response).to have_http_status 201
+      expect(Response.count).to eq 1
+
+      form_response = Response.last
+      expect(form_response.form).to eq @form
+      expect(form_response.answers.count).to eq 2
+
+      form_response.answers.each do |answer|
+        expect(answer.media_object).to be_present
+      end
```
Comment (hooverlunch) [inline spec/requests/media/odk_response_spec.rb]: Very hottttt!

File: spec/requests/media/odk_response_spec.rb [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      # Submit second part
+      post submission_path(@mission), {
+        xml_submission_file: submission_file,
+        "another_swing.jpg" => image2 },
+        "HTTP_AUTHORIZATION" => encode_credentials(@user.login, test_password)
+
+      expect(response).to have_http_status 201
+      expect(Response.count).to eq 1
+
+      form_response = Response.last
```
Comment (hooverlunch) [inline spec/requests/media/odk_response_spec.rb]: Also check odk_hash in here.

File: app/models/response.rb [hunk 15]
```diff
@@ -13,7 +13,7 @@ class Response < ActiveRecord::Base
     class_name: "Answer"
   )
 
-  attr_accessor(:modifier, :excerpts)
+  attr_accessor(:modifier, :excerpts, :awaiting_media)
```
Comment (hooverlunch) [inline app/models/response.rb]: Nice name.

**Comment by hooverlunch** [comment]
Looks great. Thanks.

**Pull request merged**